"Can't translate domain tagType.Errors.IfStuck..."

```haskell
import Clash.Prelude

-- topEntity :: Clock "System" -> Reset "System" -> Signal System Bool -> Signal System Bool
topEntity clk rst = withClockResetEnable clk rst enableGen $ mealy @System f False

f s x = (x,s)
```
In comments is the inferred type for `topEntity`.
If this is uncommented everything works correctly.

When left commented it will still type check and simulate fine in clashi.
But when generating HDL clash will trip over some leftover type magic from the multiple hidden clocks and/or the associated custom type errors: `Can't translate domain tagType.Errors.IfStuck8214565720323842538 ...`
<details>
<summary>Full error:</summary>

```
NamingClashes.hs:4:1: error:
    Clash.Netlist.Util(796): Can't translate domain tagType.Errors.IfStuck8214565720323842538
  Clash.Class.HasDomain.HasSingleDomain.TryDomainResult8214565720323844496
  GHC.Types.Symbol3674937295934325066
  (Clash.Class.HasDomain.HasSingleDomain.TryDomain8214565720323844495
     (Clash.Signal.Internal.Signal8214565720323789597
        "System"
        GHC.Types.Bool3674937295934324744
     -> Clash.Signal.Internal.Signal8214565720323789597
          "System"
          GHC.Types.Bool3674937295934324744)
     (Clash.Signal.Internal.Signal8214565720323789597
        "System"
        GHC.Types.Bool3674937295934324744
     -> Clash.Signal.Internal.Signal8214565720323789597
          "System"
          GHC.Types.Bool3674937295934324744))
  (GHC.TypeLits.TypeError3674937295934325088
     GHC.Types.Symbol3674937295934325066
     (GHC.TypeLits.:$$:3891110078048108700
        (GHC.TypeLits.:$$:3891110078048108700
           (GHC.TypeLits.:$$:3891110078048108700
              (GHC.TypeLits.:$$:3891110078048108700
                 (GHC.TypeLits.:$$:3891110078048108700
                    (GHC.TypeLits.:$$:3891110078048108700
                       (GHC.TypeLits.:$$:3891110078048108700
                          (GHC.TypeLits.:$$:3891110078048108700
                             (GHC.TypeLits.:$$:3891110078048108700
                                (GHC.TypeLits.:$$:3891110078048108700
                                   (GHC.TypeLits.:$$:3891110078048108700
                                      (GHC.TypeLits.:$$:3891110078048108700
                                         (GHC.TypeLits.:$$:3891110078048108700
                                            (GHC.TypeLits.Text3891110078048108694
                                               "Could not determine whether the following type contained a non-ambiguous domain:")
                                            (GHC.TypeLits.Text3891110078048108694
                                               ""))
                                         (GHC.TypeLits.:<>:3891110078048108697
                                            (GHC.TypeLits.Text3891110078048108694
                                               "  ")
                                            (GHC.TypeLits.ShowType3891110078048108703
                                               (GHC.Prim.TYPE3674937295934324912
                                                  GHC.Types.LiftedRep3891110078048108766)
                                               (Clash.Signal.Internal.Signal8214565720323789597
                                                  "System"
                                                  GHC.Types.Bool3674937295934324744
                                               -> Clash.Signal.Internal.Signal8214565720323789597
                                                    "System"
                                                    GHC.Types.Bool3674937295934324744))))
                                      (GHC.TypeLits.Text3891110078048108694 ""))
                                   (GHC.TypeLits.Text3891110078048108694
                                      "In the full type:"))
                                (GHC.TypeLits.Text3891110078048108694 ""))
                             (GHC.TypeLits.:<>:3891110078048108697
                                (GHC.TypeLits.Text3891110078048108694 "  ")
                                (GHC.TypeLits.ShowType3891110078048108703
                                   (GHC.Prim.TYPE3674937295934324912
                                      GHC.Types.LiftedRep3891110078048108766)
                                   (Clash.Signal.Internal.Signal8214565720323789597
                                      "System"
                                      GHC.Types.Bool3674937295934324744
                                   -> Clash.Signal.Internal.Signal8214565720323789597
                                        "System"
                                        GHC.Types.Bool3674937295934324744))))
                          (GHC.TypeLits.Text3891110078048108694 ""))
                       (GHC.TypeLits.Text3891110078048108694
                          "Does it contain one?"))
                    (GHC.TypeLits.Text3891110078048108694 ""))
                 (GHC.TypeLits.Text3891110078048108694 "------"))
              (GHC.TypeLits.Text3891110078048108694 ""))
           (GHC.TypeLits.:$$:3891110078048108700
              (GHC.TypeLits.:$$:3891110078048108700
                 (GHC.TypeLits.:$$:3891110078048108700
                    (GHC.TypeLits.:$$:3891110078048108700
                       (GHC.TypeLits.:$$:3891110078048108700
                          (GHC.TypeLits.:$$:3891110078048108700
                             (GHC.TypeLits.:$$:3891110078048108700
                                (GHC.TypeLits.Text3891110078048108694
                                   "This might happen if an instance for TryDomain is missing. Try to determine")
                                (GHC.TypeLits.Text3891110078048108694
                                   "which of the types miss an instance, and add them. Example implementations:"))
                             (GHC.TypeLits.Text3891110078048108694 ""))
                          (GHC.TypeLits.Text3891110078048108694
                             " * type instance TryDomain t (MyVector n a)    = TryDomain t a"))
                       (GHC.TypeLits.Text3891110078048108694
                          " * type instance TryDomain t (MyCircuit dom a) = Found dom"))
                    (GHC.TypeLits.Text3891110078048108694
                       " * type instance TryDomain t Terminal          = NotFound"))
                 (GHC.TypeLits.Text3891110078048108694 ""))
              (GHC.TypeLits.Text3891110078048108694
                 "Alternatively, use one of the withSpecific* functions.")))
        (GHC.TypeLits.:$$:3891110078048108700
           (GHC.TypeLits.:$$:3891110078048108700
              (GHC.TypeLits.:$$:3891110078048108700
                 (GHC.TypeLits.:$$:3891110078048108700
                    (GHC.TypeLits.:$$:3891110078048108700
                       (GHC.TypeLits.:$$:3891110078048108700
                          (GHC.TypeLits.:$$:3891110078048108700
                             (GHC.TypeLits.:$$:3891110078048108700
                                (GHC.TypeLits.:$$:3891110078048108700
                                   (GHC.TypeLits.:$$:3891110078048108700
                                      (GHC.TypeLits.:$$:3891110078048108700
                                         (GHC.TypeLits.:$$:3891110078048108700
                                            (GHC.TypeLits.:$$:3891110078048108700
                                               (GHC.TypeLits.:$$:3891110078048108700
                                                  (GHC.TypeLits.:$$:3891110078048108700
                                                     (GHC.TypeLits.:$$:3891110078048108700
                                                        (GHC.TypeLits.Text3891110078048108694
                                                           "")
                                                        (GHC.TypeLits.Text3891110078048108694
                                                           "------"))
                                                     (GHC.TypeLits.Text3891110078048108694
                                                        ""))
                                                  (GHC.TypeLits.Text3891110078048108694
                                                     "You tried to apply an explicitly routed clock, reset, or enable line"))
                                               (GHC.TypeLits.Text3891110078048108694
                                                  "to a construct with, possibly, an implicitly routed one. Clash failed to"))
                                            (GHC.TypeLits.Text3891110078048108694
                                               "unambigously determine a single domain and could therefore not route it."))
                                         (GHC.TypeLits.Text3891110078048108694
                                            "You possibly used one of these sets of functions:"))
                                      (GHC.TypeLits.Text3891110078048108694 ""))
                                   (GHC.TypeLits.Text3891110078048108694
                                      " * with{ClockResetEnable,Clock,Reset,Enable}"))
                                (GHC.TypeLits.Text3891110078048108694
                                   " * expose{ClockResetEnable,Clock,Reset,Enable}"))
                             (GHC.TypeLits.Text3891110078048108694 ""))
                          (GHC.TypeLits.Text3891110078048108694
                             "These functions are suitable for components defined over a single domain"))
                       (GHC.TypeLits.Text3891110078048108694
                          "only. If you want to use multiple domains, use the following instead:"))
                    (GHC.TypeLits.Text3891110078048108694 ""))
                 (GHC.TypeLits.Text3891110078048108694
                    " * withSpecific{ClockResetEnable,Clock,Reset,Enable}"))
              (GHC.TypeLits.Text3891110078048108694
                 " * exposeSpecific{ClockResetEnable,Clock,Reset,Enable}"))
           (GHC.TypeLits.Text3891110078048108694 ""))))
  (Fcf.Combinators.Pure8214565720323842537
     GHC.Types.Symbol3674937295934325066
     (Clash.Class.HasDomain.HasSingleDomain.ErrOnConflict8214565720323844488
        (Clash.Signal.Internal.Signal8214565720323789597
           "System"
           GHC.Types.Bool3674937295934324744
        -> Clash.Signal.Internal.Signal8214565720323789597
             "System"
             GHC.Types.Bool3674937295934324744)
        (Clash.Class.HasDomain.HasSingleDomain.TryDomain8214565720323844495
           (Clash.Signal.Internal.Signal8214565720323789597
              "System"
              GHC.Types.Bool3674937295934324744
           -> Clash.Signal.Internal.Signal8214565720323789597
                "System"
                GHC.Types.Bool3674937295934324744)
           (Clash.Signal.Internal.Signal8214565720323789597
              "System"
              GHC.Types.Bool3674937295934324744
           -> Clash.Signal.Internal.Signal8214565720323789597
                "System"
                GHC.Types.Bool3674937295934324744))))
    NB: The source location of the error is not exact, only indicative, as it is acquired after optimisations.
    The actual location of the error can be in a function that is inlined.
    To prevent inlining of those functions, annotate them with a NOINLINE pragma.
  |
4 | topEntity clk rst = withClockResetEnable clk rst enableGen $ mealy @System f False
  | ^^^^^^^^^
```
</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -50,6 +50,7 @@
   * [#895](https://github.com/clash-lang/clash-compiler/issues/895): VHDL type error when generating `Maybe (Vec 2 (Signed 8), Index 1)`
   * [#880](https://github.com/clash-lang/clash-compiler/issues/880): Custom bit representations can now be used on product types too
   * [#976](https://github.com/clash-lang/clash-compiler/issues/976): Prevent shadowing in Clash's core evaluator
+  * [#1007](https://github.com/clash-lang/clash-compiler/issues/1007): Can't translate domain tagType.Errors.IfStuck...
 
 * Fixes without issue reports:
   * Fix bug in `rnfX` defined for `Down` ([baef30e](https://github.com/clash-lang/clash-compiler/commit/baef30eae03dc02ba847ffbb8fae7f365c5287c2))

--- a/clash-ghc/src-ghc/Clash/GHC/NetlistTypes.hs
+++ b/clash-ghc/src-ghc/Clash/GHC/NetlistTypes.hs
@@ -28,7 +28,7 @@ import Clash.Core.Name                  (Name (..))
 import Clash.Core.Pretty                (showPpr)
 import Clash.Core.TyCon                 (TyConMap, tyConDataCons)
 import Clash.Core.Type
-  (LitTy (..), Type (..), TypeView (..), coreView, tyView)
+  (LitTy (..), Type (..), TypeView (..), coreView, coreView1, tyView)
 import Clash.Core.Util                  (tyNatSize, substArgTys)
 import Clash.Netlist.Util               (coreTypeToHWType, stripFiltered)
 import Clash.Netlist.Types
@@ -144,8 +144,8 @@ ghcTypeToHWType iw floatSupport = go
                [dc] -> case substArgTys dc args of
                  [_,tyView -> TyConApp _ [_,dom]] -> case tyView (coreView m dom) of
                    TyConApp _ [tag0, period0, edge0, rstKind0, init0, polarity0] -> do
-                     tag1      <- domTag tag0
-                     period1   <- domPeriod period0
+                     tag1      <- domTag m tag0
+                     period1   <- domPeriod m period0
                      edge1     <- domEdge m edge0
                      rstKind1  <- domResetKind m rstKind0
                      init1     <- domInitBehavior m init0
@@ -159,13 +159,13 @@ ghcTypeToHWType iw floatSupport = go
         "Clash.Signal.Internal.Clock"
           | [tag0] <- args
           -> do
-            tag1 <- domTag tag0
+            tag1 <- domTag m tag0
             returnN (Clock (pack tag1))
 
         "Clash.Signal.Internal.Reset"
           | [tag0] <- args
           -> do
-            tag1 <- domTag tag0
+            tag1 <- domTag m tag0
             returnN (Reset (pack tag1))
 
         "Clash.Sized.Internal.BitVector.Bit" -> returnN Bit
@@ -265,13 +265,15 @@ liftE
   -> ExceptT e (MaybeT m) a
 liftE = mapExceptT (MaybeT . pure . Just . coerce)
 
-domTag :: Monad m => Type -> ExceptT String (MaybeT m) String
-domTag (LitTy (SymTy tag)) = pure tag
-domTag ty = throwE $ "Can't translate domain tag" ++ showPpr ty
+domTag :: Monad m => TyConMap -> Type -> ExceptT String (MaybeT m) String
+domTag m (coreView1 m -> Just ty) = domTag m ty
+domTag _ (LitTy (SymTy tag)) = pure tag
+domTag _ ty = throwE $ "Internal error. Cannot translate domain tag:\n" ++ showPpr ty
 
-domPeriod :: Monad m => Type -> ExceptT String (MaybeT m) Integer
-domPeriod (LitTy (NumTy period)) = pure period
-domPeriod ty = throwE $ "Can't translate domain period" ++ showPpr ty
+domPeriod :: Monad m => TyConMap -> Type -> ExceptT String (MaybeT m) Integer
+domPeriod m (coreView1 m -> Just ty) = domPeriod m ty
+domPeriod _ (LitTy (NumTy period)) = pure period
+domPeriod _ ty = throwE $ "Internal error. Cannot translate domain period:\n" ++ showPpr ty
 
 fromType
   :: Monad m

--- a/tests/shouldwork/Signal/T1007.hs
+++ b/tests/shouldwork/Signal/T1007.hs
@@ -1,0 +1,7 @@
+module T1007 where
+
+import Clash.Prelude
+
+topEntity clk rst = withClockResetEnable clk rst enableGen $ mealy @System f False
+
+f s x = (x,s)

--- a/testsuite/Main.hs
+++ b/testsuite/Main.hs
@@ -340,6 +340,7 @@ runClashTest = defaultMain $ clashTestRoot
           ]
 
         , runFailingTest ("tests" </> "shouldfail" </> "Signal") allTargets [] "MAC" (Just "Couldn't instantiate blackbox for Clash.Signal.Internal.register#")
+        , runTest "T1007" def{hdlSim=False}
         ]
       , clashTestGroup "SimIO"
         [ runTest "Test00" def {hdlTargets=[Verilog], stdErrEmptyFail=False, topEntity=TopEntity "topEntity"}

